File: code/cpp/pointers.cpp
#include <memory>

using namespace std;

#define unused __attribute ((unused))
int main(unused int argc, unused char **argv)
// BEGIN_CODE
{
    // memory freed when ptr_a goes out of scope
    unique_ptr<char> ptr_a(new char('a'));

    // reference counted pointer
    shared_ptr<char> ptr_b(new char('b'));

    // pointer that does not increase reference counts
    weak_ptr<char> ptr_c = ptr_b;

    // memory freed when ptr_d goes out of scope
    // differs from unique_ptr in copy semantics
    auto_ptr<char> ptr_d(new char('d'));
}
// END_CODE

File: code/rust/pointers.rs
use std::rc::Rc;
use std::sync::Arc;

#[allow(unused_variables)]
fn main()
// BEGIN_CODE
{
    // memory freed when ptr_a goes out of scope
    let ptr_a = Box::new('a');

    // reference counted pointer
    let ptr_b = Rc::new('b');

    // pointer that does not increase reference counts
    let ptr_c = ptr_b.downgrade();

    // atomically reference counted pointer
    // can be safely used across threads
    let ptr_e = Arc::new('e');
}
// END_CODE

